fix!: change consumer quotas variable format to prevent "Duplicate object key" TF Error

mmajor-rtbh opened the pull request:
This is a **breaking change** as it changes the way TF sees it's resources.

The reason of this change is simple, currently when you create a config of consumer_quotas like this:
```  consumer_quotas             = [
    {
      service = "bigquery.googleapis.com"
      metric  = urlencode("bigquery.googleapis.com/quota/query/usage")
      limit   = urlencode("/d/project")
      dimensions = {}
      value = "100000"
    },    {
      service = "bigquery.googleapis.com"
      metric  = urlencode("bigquery.googleapis.com/quota/query/usage")
      limit   = urlencode("/d/project/user")
      dimensions = {}
      value = "100000"
    }
    ]
```
So you want to override quota both for project and project/user limit, then after you run `terrafrom plan` you get an error:
```
  │ Error: Duplicate object key
  │ 
  │   on .terraform/modules/projects/modules/quota_manager/main.tf line 18, in locals:
  │   18:   consumer_quotas = { for index, quota in var.consumer_quotas : "${quota.service}-${quota.metric}" => quota }
  │     ├────────────────
  │     │ quota.metric is "bigquery.googleapis.com%2Fquota%2Fquery%2Fusage"
  │     │ quota.service is "bigquery.googleapis.com"
  │ 
  │ Two different items produced the key
  │ "bigquery.googleapis.com-bigquery.googleapis.com%2Fquota%2Fquery%2Fusage"
  │ in this 'for' expression. If duplicates are expected, use the ellipsis
  │ (...) after the value expression to enable grouping by key.
```

This change fixes this issue and potentially other similar when you want to set up different limit for same quota/metric.

Proposed change is to change variable name from "quota-metric" value to "metric-limit" not for "qouta-metric-limit" as metric always consists of quota value plus quota name, ie:
quota = "bigquery.googleapis.com"
metric = "bigquery.googleapis.com%2Fquota%2Fquery%2Fusage"

mmajor-rtbh committed:
Update consumer quotas local variable format

mmajor-rtbh committed:
Merge branch 'terraform-google-modules:main' into change_consumer_quotas_variable_format

bharathkkb reviewed (commented):
Thanks for the PR @rtb-major

mmajor-rtbh committed:
Merge branch 'terraform-google-modules:main' into change_consumer_quotas_variable_format

mmajor-rtbh committed:
Update modules/quota_manager/main.tf

I'am not going argue about that, personaly I like simpler solutions. But maybe there will be some "edge case" where my proposition won't be working

Co-authored-by: Bharath KKB <bharathkrishnakb@gmail.com>

mmajor-rtbh committed:
Merge branch 'terraform-google-modules:main' into change_consumer_quotas_variable_format

mmajor-rtbh committed:
Merge branch 'terraform-google-modules:main' into change_consumer_quotas_variable_format

mmajor-rtbh commented:
hey, 5 months has passed and no one can look into this, is this repo live or not ?

bharathkkb committed:
Merge branch 'main' into change_consumer_quotas_variable_format

bharathkkb commented:
/gcbrun

mmajor-rtbh commented:
bump

mmajor-rtbh commented:
bump

daniel-cit commented:
/gcbrun

mmajor-rtbh commented:
do I need to fix something or why is cicd failing ?

mmajor-rtbh committed:
Merge branch 'main' into change_consumer_quotas_variable_format
